price-estimator/src/orderbook.rs

```diff
@@ -0,0 +1,128 @@
+};
+use pricegraph::{Element, Price, TokenPair, UserId, Validity};
+use primitive_types::U256;
+use std::time::{Duration, SystemTime};
+use tokio::sync::RwLock;
+
+/// Access and update the pricegraph orderbook.
+pub struct Orderbook<T> {
+    orderbook_reading: T,
+    reduced_pricegraph_orderbook: RwLock<pricegraph::Orderbook>,
+}
+
+impl<T> Orderbook<T> {
+    pub fn new(orderbook_reading: T) -> Self {
+        Self {
+            orderbook_reading,
+            reduced_pricegraph_orderbook: RwLock::new(pricegraph::Orderbook::from_elements(
+                std::iter::empty(),
+            )),
+        }
+    }
+
+    pub async fn _get_reduced_orderbook(&self) -> pricegraph::Orderbook {
+        self.reduced_pricegraph_orderbook.read().await.clone()
+    }
+}
+
+impl<T: StableXOrderBookReading> Orderbook<T> {
+    /// Recreate the pricegraph orderbook.
+    pub async fn update(&self) -> anyhow::Result<()> {
+        let (account_state, orders) = self
+            .orderbook_reading
+            .get_auction_data(current_batch_id() as u32)
+            .await?;
+
+        // TODO: Move this cpu heavy computation out of the async function using spawn_blocking.
+        let mut orderbook = pricegraph::Orderbook::from_elements(
+            orders
+                .iter()
+                .map(|order| order_to_element(order, &account_state)),
+        );
+        orderbook.reduce_overlapping_orders();
+
+        *self.reduced_pricegraph_orderbook.write().await = orderbook;
+        Ok(())
+    }
+}
+
+/// Current batch id based on system time.
+fn current_batch_id() -> u64 {
+    const BATCH_DURATION: Duration = Duration::from_secs(300);
+    let now = SystemTime::now();
+    let time_since_epoch = now
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("unix epoch is not in the past");
+    time_since_epoch.as_secs() / BATCH_DURATION.as_secs()
+}
+
+/// Convert a core Order to a pricegraph Element.
+fn order_to_element(order: &Order, account_state: &AccountState) -> Element {
+    // Some conversions are needed because the primitive types crate is on different versions in
+    // core and pricegraph.
+    Element {
+        user: UserId::from_slice(order.account_id.as_fixed_bytes()),
+        balance: U256(
+            account_state
+                .read_balance(order.sell_token, order.account_id)
+                .0,
+        ),
+        pair: TokenPair {
+            buy: order.buy_token,
+            sell: order.sell_token,
+        },
+        valid: Validity {
+            from: order.valid_from,
+            to: order.valid_until,
+        },
+        price: Price {
+            numerator: order.numerator,
+            denominator: order.denominator,
+        },
+        remaining_sell_amount: order.remaining_sell_amount,
+        id: order.id,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ethcontract::{Address, U256};
+    use std::collections::HashMap;
+
+    #[test]
+    fn order_to_element_() {
```

Comment:
ubernit: Not of fan of the `method_` naming, maybe the test name could describe what the test is doing like:
`order_is_converted_to_auction_element`.

Reply:
Or, would it be a reasonable compromise to use `test_method_name`? I believe this is a common convention.

Reply:
The reason I use this name is that we are testing that whole function. I don't see how another name is helpful. I would use a more detailed name if we had more test cases like `convert_order_large_number`. Likewise prepending `test_` doesn't add any information because it is already clear that this is a test.

Reply:
> The reason I use this name is that we are testing that whole function. I don't see how another name is helpful. I would use a more detailed name if we had more test cases like `convert_order_large_number`.

My main argument is that it is not necessarily testing the whole function, but a single aspect of it. Additional tests (like `convert_order_large_number`) will be testing other aspects of this function, and if and when they are added, the original test is no longer testing the whole function and is no longer accurately named. This could happen if some refactoring or new code introduces new branches that need to be tested and deserve their own unit tests. As the code get modified, the original goal of the unit test might become obscured. Furthermore, a refactor might split the method into smaller pieces, in which case this unit test will be for a new method.

That being said, for this method I think the way it currently is is fine. It really was an **uber**nit and I don't think it necessarily needs to change.

> Likewise prepending test_ doesn't add any information because it is already clear that this is a test.

I also agree with this, especially since the function is already in a tests module so it shows up as `tests::method_name` and have a clear `#[test]` attribute.